ethereum/evmtool/src/main/java/org/hyperledger/besu/evmtool/T8nSubCommand.java

```diff
@@ -0,0 +1,557 @@
+        }
+        final Account txSender = worldStateUpdater.getAccount(transaction.getSender());
+        if (txSender != null && txSender.isEmpty()) {
+          worldStateUpdater.deleteAccount(txSender.getAddress());
+        }
+      }
+      if (result.isInvalid()) {
+        invalidTransactions.add(
+            new RejectedTransaction(i, result.getValidationResult().getErrorMessage()));
+      } else {
+        validTransactions.add(transaction);
+
+        long transactionGasUsed = transaction.getGasLimit() - result.getGasRemaining();
+
+        gasUsed += transactionGasUsed;
+        long intrinsicGas =
+            protocolSpec
+                .getGasCalculator()
+                .transactionIntrinsicGasCost(
+                    transaction.getPayload(), transaction.getTo().isEmpty());
+        tracer.traceEndTransaction(
+            result.getOutput(), gasUsed - intrinsicGas, timer.elapsed(TimeUnit.NANOSECONDS));
+        TransactionReceipt receipt =
+            protocolSpec
+                .getTransactionReceiptFactory()
+                .create(transaction.getType(), result, worldState, gasUsed);
+        Bytes gasUsedInTransaction = Bytes.ofUnsignedLong(transactionGasUsed);
+        receipts.add(receipt);
+        ObjectNode receiptObject = receiptsArray.addObject();
+        receiptObject.put(
+            "root", receipt.getStateRoot() == null ? "0x" : receipt.getStateRoot().toHexString());
+        receiptObject.put("status", "0x" + receipt.getStatus());
+        receiptObject.put("cumulativeGasUsed", Bytes.ofUnsignedLong(gasUsed).toQuantityHexString());
+        receiptObject.put("logsBloom", receipt.getBloomFilter().toHexString());
+        if (result.getLogs().isEmpty()) {
+          receiptObject.putNull("logs");
+        } else {
+          ArrayNode logsArray = receiptObject.putArray("logs");
+          for (Log log : result.getLogs()) {
+            logsArray.addPOJO(log);
+          }
+        }
+        receiptObject.put("transactionHash", transaction.getHash().toHexString());
+        receiptObject.put(
+            "contractAddress", transaction.contractAddress().orElse(Address.ZERO).toHexString());
+        receiptObject.put("gasUsed", gasUsedInTransaction.toQuantityHexString());
+        receiptObject.put("blockHash", Hash.ZERO.toHexString());
+        receiptObject.put("transactionIndex", Bytes.ofUnsignedLong(i).toQuantityHexString());
+      }
+    }
+
+    final ObjectNode resultObject = objectMapper.createObjectNode();
+
+    // block reward
+    if (!validTransactions.isEmpty()) {
+      worldStateUpdater
+          .getOrCreateSenderAccount(blockHeader.getCoinbase())
+          .getMutable()
+          .incrementBalance((reward == null) ? protocolSpec.getBlockReward() : reward);
+    }
+
+    // deposit CL withdrawals
+    try {
+      protocolSpec
+          .getWithdrawalsProcessor()
+          .ifPresent(
+              p -> p.processWithdrawals(referenceTestEnv.getWithdrawals(), worldStateUpdater));
```

Comment:
I might be misunderstanding the point of this tool, but aren't we duplicating and potentially creating diverging behaviour from how mainnet besu behaves?

For example, in AbstractBlockProcessor, we only process withdrawals if both the processor is present and also if withdrawals are present, see: https://github.com/hyperledger/besu/blob/main/ethereum/core/src/main/java/org/hyperledger/besu/ethereum/mainnet/AbstractBlockProcessor.java#L143-L149

Same applies to other operations conditionally invoked by this tool. At a glance, it seems like behaviour of mainnet vs this tool may diverge further over time too, so doesn't that invalidate the test?

Reply:
They should be the same.  But the difficulty arises in how the test needs to instrument it's output, in addition to other presumptions about how to handle invalid transactions (abort whole block in production vs. discard and produce some block output in test).

This tool also, as much as is exposed, delegates to other existing components to calculate values.

This tool in itself is not the test, but generates the testing vectors.  Because of that it needs to create some output from testing inputs that would ordinarily be fatal to regular processing.  That's why simply sending it off to a block processor won't work because it's not enough to know it fails, but to produce certain vectors for specific invalid calls.

Reply:
Thanks for the explanation. Wonder if there's a way to instrument in line in the code? Maybe not worth the readability/complexity cost of doing so though.

Reply:
And we would have to factor out block production, which is currently baked into the AbstractBlockProcessor.

One alternative is a static method that is public and factored out where you bring all the pieces needed to process a block, and have both tools call that at the appropriate time and place.